// imagekitio-angular/esm2015/lib/imagekitio-angular.service.js
import { Injectable } from '@angular/core';
import ImageKit from 'imagekit-javascript';
import * as i0 from "@angular/core";
const SDK_VERSION = '5.0.0';
export class ImageKitConfiguration {
}
export class ImagekitService {
    constructor(configuration) {
        this.configuration = configuration;
        (configuration.sdkVersion = `angular-${SDK_VERSION}`),
            (this._ikInstance = new ImageKit(this.configuration));
    }
    get ikInstance() {
        return this._ikInstance;
    }
    getUrl(config) {
        const url = this._ikInstance.url(config);
        return url;
    }
}
ImagekitService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: ImagekitService, deps: [{ token: ImageKitConfiguration }], target: i0.ɵɵFactoryTarget.Injectable });
ImagekitService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: ImagekitService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: ImagekitService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: ImageKitConfiguration }]; } });
