// imagekitio-angular/esm2015/lib/imagekitio-angular.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { IkImageComponent } from './ik-image/ik-image.component';
import { IkUploadComponent } from './ik-upload/ik-upload.component';
import { IkVideoComponent } from './ik-video/ik-video.component';
import { ImageKitConfiguration, ImagekitService } from './imagekitio-angular.service';
import * as i0 from "@angular/core";
export class ImagekitioAngularModule {
    static forRoot(config) {
        return {
            ngModule: ImagekitioAngularModule,
            providers: [
                { provide: ImageKitConfiguration, useValue: config }
            ]
        };
    }
}
ImagekitioAngularModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: ImagekitioAngularModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ImagekitioAngularModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: ImagekitioAngularModule, declarations: [IkImageComponent,
        IkVideoComponent,
        IkUploadComponent], imports: [CommonModule], exports: [IkImageComponent,
        IkVideoComponent,
        IkUploadComponent] });
ImagekitioAngularModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: ImagekitioAngularModule, providers: [ImagekitService], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: ImagekitioAngularModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        IkImageComponent,
                        IkVideoComponent,
                        IkUploadComponent
                    ],
                    imports: [CommonModule],
                    exports: [
                        IkImageComponent,
                        IkVideoComponent,
                        IkUploadComponent
                    ],
                    providers: [ImagekitService]
                }]
        }] });
